File: PDOO/NapakalakiRuby.ruby/lib/Napakalaki.rb
#encoding: utf-8

require "singleton"
require_relative "card_dealer"
require_relative "player"
require_relative 'combat_result'
require_relative 'cultist_player'


module NapakalakiGame
  
  class Napakalaki
    
    include Singleton
    
    private
    
    def initialize
      @currentMonster = nil
      @currentPlayer = nil
      @players = Array.new
      @dealer = CardDealer.instance 
    end
    
    def initPlayers(names)
      for name in names
        @players << Player.new(name)
      end
    end
    
    def nextPlayer
      if @currentPlayer == nil
        @currentPlayer = @players.at(rand(@players.size))
      else
        indice = @players.index(@currentPlayer)
        if (@players.size() - 1) == indice
          indice = 0
        else
          indice += 1
        end
        @currentPlayer = @players.at(indice)
      end 
      @currentPlayer
    end
    
    def nextTurnAllowed
      allowed = false;
      allowed = (@currentPlayer == nil) || @currentPlayer.validState
    end
    
    def setEnemies
      enemigo = @players.at(0)
      for player in @players
        loop do 
         enemigo = @players.at(rand(@players.size))
        break if (player != enemigo)
        end
        player.setEnemy(enemigo)
      end
      
    end
    
    public 
    
    def developCombat
      combatResult = @currentPlayer.combat(@currentMonster)
      @dealer.giveMonsterBack(@currentMonster)
      
      if (combatResult == CombatResult::LOSEANDCONVERT)
        card = CardDealer.instance.nextCultist
        sec = CultistPlayer.new(@currentPlayer, card)
        @players.each do |p|
          if (p.getEnemy == @currentPlayer)
            p.setEnemy(sec)
          end
        end
        i = @players.find_index(@currentPlayer)
        @players[i] = sec
        @currentPlayer = sec
      end
      
      return combatResult
    end
    
    def discardVisibleTreasures(treasures)
      for treasure in treasures
        @currentPlayer.discardVisibleTreasure(treasure)
        @dealer.giveTreasureBack(treasure)
      end
    end
    
    def discardHiddenTreasures(treasures)
      for treasure in treasures
        @currentPlayer.discardHiddenTreasure(treasure)
        @dealer.giveTreasureBack(treasure)
      end
    end
    
    def makeTreasuresVisible(treasures)
      for treasure in treasures
        @currentPlayer.makeTreasureVisible(treasure)
      end
    end
    
    def initGame(players)
      initPlayers(players)
      setEnemies
      @dealer.initCards
      nextTurn
    end
    
    def getCurrentPlayer
      @currentPlayer
    end
    
    def getCurrentMonster
      @currentMonster
    end
    
    def nextTurn
      stateOK = nextTurnAllowed
      if (stateOK)
        @currentMonster = @dealer.nextMonster
        @currentPlayer = nextPlayer
        dead = @currentPlayer.isDead
        
        if (dead)
          @currentPlayer.initTreasures
        end
      end
      stateOK
    end
   
    def endOfGame(result)
      res = result == CombatResult::WINGAME
      return res
    end
  end
end

File: PDOO/Napakalaki.java/src/NapakalakiGame/Napakalaki.java
package NapakalakiGame;

/**
 *
 * @author Andres Arco Lopez y Miguel Angel Cantarero Lopez
 */

import java.util.ArrayList;
import java.util.Random;

public class Napakalaki {
    
    private static final Napakalaki instance = new Napakalaki(); 
    private Monster currentMonster;
    private CardDealer dealer = CardDealer.getInstance();
    public Player currentPlayer;
    private ArrayList<Player> players;
    
    private Napakalaki(){
        currentMonster = null;
        currentPlayer = null;
        players = null;
    }
    
    public static Napakalaki getInstance(){
        return instance;
    }
    
    private void initPlayers(ArrayList <String> names){
        players = new ArrayList();
        
        for (String s : names)
            players.add(new Player(s));
                    
    }
    
    private Player nextPlayer(){
        int jugador;
        if (currentPlayer == null){
            jugador = (int) (Math.random() *(players.size()));
            currentPlayer = players.get(jugador);
        }
        else{
            jugador = players.indexOf(currentPlayer)+1;
            if(jugador == players.size()){
                currentPlayer = players.get(0);
            }
                else {
               currentPlayer = players.get(jugador);
        }
    }
        return currentPlayer;
    }
    
    private boolean nextTurnAllowed(){
        boolean allowed;
        if (currentPlayer == null){
            allowed = true;
        }
        else{
            allowed = currentPlayer.validState();
        }
        
        return allowed;
    }
    
    private void setEnemies(){
        int jugador;
        for(int i = 0; i < players.size() ; ++i){
            jugador = (int) (Math.random() *(players.size()));
            while(jugador == i){
                jugador = (int) (Math.random() *(players.size()));
            }
            players.get(i).setEnemy(players.get(jugador));
        
        }
    }
    
    public CombatResult developCombat(){
        CombatResult combatResult = currentPlayer.combat(currentMonster);
        
        if(combatResult==CombatResult.LOSEANDCONVERT){
            CultistPlayer sectario = new CultistPlayer(currentPlayer, CardDealer.getInstance().nextCultist());
            
            for(Player p:players){
                if(p.getEnemy()==currentPlayer)
                    p.setEnemy(sectario);
            }
        
            players.set(players.indexOf(currentPlayer),sectario);
            currentPlayer = sectario;
        }
        
        CardDealer.getInstance().giveMonsterBack(currentMonster);
        return combatResult;
    }
    
    public void discardVisibleTreasures(ArrayList <Treasure> treasures){
        for(Treasure treasure : treasures){
            currentPlayer.discardVisibleTreasure(treasure);
            dealer.giveTreasureBack(treasure);
        }
    }
    
    public void discardHiddenTreasures(ArrayList <Treasure> treasures){
        for(Treasure treasure : treasures){
            currentPlayer.discardHiddenTreasure(treasure);
            dealer.giveTreasureBack(treasure);
    }
    }
    
    public void makeTreasuresVisible(ArrayList <Treasure> treasures){
        for(Treasure t : treasures){
            currentPlayer.makeTreasureVisible(t);            
        }
    }
    
    public void initGame(ArrayList <String> players){
        initPlayers(players);
        setEnemies();
        dealer.initCards();
        nextTurn();
    }
    
    public Player getCurrentPlayer(){
        return currentPlayer;
    }
    
    public Monster getCurrentMonster(){
        return currentMonster;
    }
    
    public boolean nextTurn(){
       
        boolean stateOK = nextTurnAllowed();
        
        if(stateOK){
            currentMonster = dealer.nextMonster();
            currentPlayer = nextPlayer();
            boolean dead = currentPlayer.isDead();
            if(dead){
                currentPlayer.initTreasures();
            }
        }
        return stateOK;
            
    }
    
    public boolean endOfGame(CombatResult result){
        boolean eog = false;
        if (result == CombatResult.WINGAME){
            eog = true;
        }
        return eog;     
            
    }
    
}

